### PR TITLE
feat: Drag to Reorder and Resize Experiment Columns columns [DET-6438] [DET-6809]

### PR DESCRIPTION
**Pull request by trentwatt**
## Description

feat: Drag to Reorder and Resize Experiment Columns columns [DET-6438] [DET-6809]

## Commentary (optional)

Separated out the new functionality into a separate table component (apart from ResponsiveTable), along with the custom context menu functionality (which was then removed from ResponsiveTable) 

the main pieces of data are 
* settings.columns the currently displayed columns and their ordering
* settings.columnWidths the most recently 'saved' widths
* widthData.widths is used internally by the table for rendering e.g. during resize. settings.columnWidths is not updated until a resize finishes or until a width adjustment comes from outside.

when the sum of widths of the table columns is less than the table container, the table expands to fill the container. then one 'pixel' of column width corresponds to more than one pixel on the screen. this messes up resize tracking, which assumes 1px of mouse drag corresponds to 1px of resizing. the solution is to upscale settings.columnWidths so that they fill the table. this extends to doing so during resize as well. if a resize brings the sum of widths to be less than the width of the table, we would lose tracking then as well. the solution in that case is to distribute the deficit equally among widthData.widths that are not being resized. 

To keep track of the container width, we have to drop a ref into the page component and then useResize on it from within the table.

## Checklist

- [x] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](../docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

[DET-6438]: https://determinedai.atlassian.net/browse/DET-6438?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[DET-6809]: https://determinedai.atlassian.net/browse/DET-6809?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: webui/react/src/pages/ExperimentList.tsx [hunk 1]
```diff
@@ -308,7 +315,49 @@ const ExperimentList: React.FC = () => {
...
+        align: 'right',
+        className: 'fullCell',
+        fixed: 'right',
+        key: 'action',
+        onCell: () => ({ isCellRightClickable: true } as React.HTMLAttributes<HTMLElement>),
+        render: actionRenderer,
+        title: '',
+        width: DEFAULT_COLUMN_WIDTHS['action' as ExperimentColumnName],
+      },
+      archived:
```
Comment (EmilyBonar) [inline webui/react/src/pages/ExperimentList.tsx:330]: chore: I'm not sure what's causing this, but toggling the archived slider isn't working on Netlify. Once the archived columns appears, it won't disappear.

Comment (EmilyBonar) [reply webui/react/src/pages/ExperimentList.tsx:330]: Archived also shouldn't be an option in the column picker modal. This is probably related.

File: webui/react/src/pages/ExperimentList.tsx [hunk 2]
```diff
@@ -308,7 +315,49 @@ const ExperimentList: React.FC = () => {
...
+      action:
+      {
+        align: 'right',
+        className: 'fullCell',
+        fixed: 'right',
+        key: 'action',
+        onCell: () => ({ isCellRightClickable: true } as React.HTMLAttributes<HTMLElement>),
+        render: actionRenderer,
+        title: '',
+        width: DEFAULT_COLUMN_WIDTHS['action' as ExperimentColumnName],
```
Comment (EmilyBonar) [inline webui/react/src/pages/ExperimentList.tsx:328]: chore: This causes weird behavior when you try to resize the action column.

File: webui/react/src/pages/ExperimentList.tsx [hunk 3]
```diff
@@ -564,23 +571,55 @@ const ExperimentList: React.FC = () => {
...
+      if (!settings.columns?.length || !settings.columnWidths?.length) return;
+      const columns = [ ...settings.columns, 'archived' ] as ExperimentColumnName[];
+      const columnWidths = [ ...settings.columnWidths, DEFAULT_COLUMN_WIDTHS['archived'] ];
+      updateSettings({
+        archived: showArchived,
+        columns: columns,
+        columnWidths: columnWidths,
+        row: undefined,
+      });
+    }
```
Comment (EmilyBonar) [inline webui/react/src/pages/ExperimentList.tsx:620]: Oh, this is probably where the archived toggle is breaking. Not 100% sure which line of code is causing it, but it's definitely broken.

File: webui/react/src/components/InteractiveTable.tsx [hunk 4]
```diff
@@ -0,0 +1,510 @@
...
+  const [ , drag ] = useDrag({
+    canDrag: () => !isResizing,
+    item: { index },
+    type,
+  });
+
+  const [ { isOver }, drop ] = useDrop({
+    accept: type,
+    canDrop: (_, monitor) => {
+      const dragItem = (monitor.getItem() || {}); // as DndItem;
```
Comment (EmilyBonar) [inline webui/react/src/components/InteractiveTable.tsx:188]: nit: either uncomment the type assertion or remove it

**Review by EmilyBonar — CHANGES_REQUESTED** [review]
Mostly looks good! A couple of things still need to be fixed though, mainly InteractiveTable's reliance on Experiment List specific values and the behavior of the archived toggle.

**Comment by trentwatt** [comment]
Made the changes, and also kept the Table API more consistent with regard to columns as we discussed. 

Will create a ticket to address any further UX improvements: stickyness for buttons, edge column resizing, etc.

File: webui/react/src/components/InteractiveTable.tsx [hunk 5]
```diff
@@ -0,0 +1,510 @@
...
+      />
+      <DraggableCore
+        nodeRef={resizingRef}
+        onDrag={onResize}
+        onStart={onResizeStart}
+        onStop={onResizeStop}>
+        <span
+          className={css.columnResizeHandle}
+          ref={resizingRef}
+          onClick={(e) => {
```
[the remark this answers is not in the document]
Comment (trentwatt) [reply webui/react/src/components/InteractiveTable.tsx]: made change. left as inline function, but I went ahead and defined `onRightClickableCell` above in `ExperimentList.tsx` based on this feedback.

File: webui/react/src/components/InteractiveTable.tsx [hunk 6]
```diff
@@ -0,0 +1,510 @@
...
+      }
+      return newWidths.map(Math.round);
+    },
+    [ pageWidth, getAdjustedColumnWidthSum ],
+  );
+
+  useLayoutEffect(() => {
+    const newSettingsWidths = settings.columnWidths;
+    const widths = getUpscaledWidths(newSettingsWidths);
+    const c = 0.95;
```
[the remark this answers is not in the document]
Comment (trentwatt) [reply webui/react/src/components/InteractiveTable.tsx]: this was just to make the drop targets smaller/more distinct for testing purposes, I took it out.

File: webui/react/src/components/InteractiveTable.tsx [hunk 7]
```diff
@@ -0,0 +1,510 @@
...
+            targetWidths = prevWidths.map((w: number, i: number) =>
+              index === i ? newWidth : w);
+          }
+
+          const targetWidthSum = getAdjustedColumnWidthSum(targetWidths);
+
+          const shortage = pageWidth - targetWidthSum;
+          if (shortage > 0) {
+            const compensatingPortion = shortage / (prevWidths.length - 1);
+            targetWidths = targetWidths.map((w, i) => index === i ? w : w + compensatingPortion);
```
[the remark this answers is not in the document]
Comment (trentwatt) [reply webui/react/src/components/InteractiveTable.tsx]: I made naming improvements throughout this section, I think it reads a lot better now.

File: webui/react/src/components/InteractiveTable.tsx [hunk 8]
```diff
@@ -0,0 +1,510 @@
...
+  useState,
+} from 'react';
+import 'antd/dist/antd.min.css';
+import { useDrag, useDragLayer, useDrop } from 'react-dnd';
+import { DraggableCore, DraggableData, DraggableEventHandler } from 'react-draggable';
+import { throttle } from 'throttle-debounce';
+
+import useResize from 'hooks/useResize';
+import { DEFAULT_COLUMN_WIDTHS, Settings } from 'pages/ExperimentList.settings';
+import { ExperimentItem } from 'types';
```
[the remark this answers is not in the document]
Comment (trentwatt) [reply webui/react/src/components/InteractiveTable.tsx]: Ok, this is fixed.

File: webui/react/src/pages/ExperimentList.tsx (hunk 2 above)
Comment (trentwatt) [reply webui/react/src/pages/ExperimentList.tsx]: To be addressed in future UX refinement

File: webui/react/src/pages/ExperimentList.tsx (hunk 3 above)
Comment (trentwatt) [reply webui/react/src/pages/ExperimentList.tsx]: I have restored the original behavior of archived toggle/column

File: webui/react/src/components/InteractiveTable.tsx (hunk 4 above)
Comment (trentwatt) [reply webui/react/src/components/InteractiveTable.tsx]: done.

**Review by EmilyBonar — APPROVED** [review]
LGTM!

**Pull request merged**